"Rent" larger program size

shiqizng opened the pull request:
Allow paying for more program space

jannotti commented on data/transactions/transaction.go:
```diff
@@ -359,6 +359,10 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 			return fmt.Errorf("tx.GlobalStateSchema too large, max number of keys is %d", proto.MaxGlobalSchemaEntries)
 		}
 
+		if tx.ExtraProgramPages > proto.MaxExtraAppProgramPages {
```
Very small complaint - seem like we ought to move this check up, so that tx.ExtraProgramPages is checked before using it on lines 346 and 350.  No difference in result, but it seem right.

jannotti reviewed (commented):
This is looking really good so far. I mentioned a small unit test that can happen for the min_balance opcode, I don't know how many unit tests we have a txns - I expect you'll most need to do most checks as integration tests.

algorandskiy commented on data/transactions/transaction.go:
```diff
@@ -343,12 +343,12 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 
-		if len(tx.ClearStateProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("clear state program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if len(tx.ClearStateProgram) > ((1 + tx.ExtraProgramPages) * proto.MaxAppProgramLen) {
+			return fmt.Errorf("clear state program too long. max len %d bytes", (1+tx.ExtraProgramPages)*proto.MaxAppProgramLen)
```
are we allowing a single page of `ExtraProgramPages` to be count towards both clear state and approval? One can pay for a page but utilize 2x space

jannotti replied:
With these changes, paying extra will get you an extra 1k for each.  I intend to add changes that combine them later, and then each page will get to 2k total extra.

algorandskiy commented on ledger/apply/application.go:
```diff
@@ -114,6 +115,11 @@ func createApplication(ac *transactions.ApplicationCallTxnFields, balances Balan
 
+	// Update the cached TotalExtraAppPages for this account, used
+	// when computing MinBalance
+	totalExtraPages := record.TotalExtraAppPages
+	totalExtraPages += ac.ExtraProgramPages
```
these variables are not used, and `TotalExtraAppPages` is not assigned

algorandskiy commented on data/transactions/transaction.go:
```diff
@@ -343,12 +343,12 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 
-		if len(tx.ApprovalProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("approval program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if len(tx.ApprovalProgram) > ((1 + tx.ExtraProgramPages) * proto.MaxAppProgramLen) {
```
usage of `tx.ExtraProgramPages` needs to be under a feature switch. Consider a scenario:
1. Half of nodes upgraded to `ExtraProgramPages`, other half is not.
2. Nodes receive a txn with `ExtraProgramPages` = 1 and `len(ApprovalProgram) = proto.MaxAppProgramLen * 1.5`. Then upgrade nodes will pass the txn even before the protocol upgrade, and other will break (most likely on txn decoding)

jannotti replied:
Until the protocol upgrade, having a ExtraProgramPages > 0 will be rejected either way.  An upgraded node will be mad that ExtraProgramPages > 0 (since proto.MaxExtraAppProgramPages will be initialized to 0), while an unupgraded node will be made it contains the field at all. So I think the protocol upgrade itself acts as the feature flag.

ExtraProgramPages = 0 would be trouble, except we will reject that because it should be omitted when 0.

algorandskiy replied:
correct if moving `if tx.ExtraProgramPages > proto.MaxExtraAppProgramPages` check right above.
I assume old code will fail on decoding `ExtraProgramPages` but it is better to double check.

algorandskiy replied:
> ExtraProgramPages = 0 would be trouble, except we will reject that because it should be omitted when 0.

@jannotti could you elaborate? 0 looks ok here.

jannotti replied:
We usually use 'omitempty' (and certainly should here, for the extra pages parameter) and when we do that, I think it becomes impossible to sign a transaction with 0 in it for that field, because we must do some check where we reencode to canonical representation (which will differ, since the 0 field will be dropped) and check the signature against that.

shiqizng replied:
in that case, a program of len <= proto.MaxAppProgramLen will get dropped?

jannotti replied:
No, I was just pointing out that @algorandskiy 's concern might occur if, before the protocol upgrade:
a) Node with new release accepts a txn containing extrapages=0
b) Not upgraded node can't understand it.

But, that's not actually possible.  Txns can't *explicitly* set the extra pages field to 0. Instead, they omit it if they mean 0. So I don't think anything bad can happen.

algorandskiy replied:
still I think we are fine:
1. old node receives a txn with omitted `ExtraProgramPages` => it decodes and proceeds as now
2. new node creates a txn without `ExtraProgramPages` => fine
3. new node receives a txn with omitted`ExtraProgramPages` => decodes as ExtraProgramPages=0 and proceeds with new code checks
4. new node receives a txn with non-zero `ExtraProgramPages` => fine, execute new code checks

jannotti replied:
I agree

algorandskiy replied:
I agree. So move the check `tx.ExtraProgramPages > proto.MaxExtraAppProgramPages` above and write a test checking
1) pre-upgrade txn with ExtraProgramPages=1 fails
2) post-upgrade txn with ExtraProgramPages=1 succeed

jannotti commented on data/transactions/application.go:
```diff
@@ -134,14 +134,19 @@ type ApplicationCallTxnFields struct {
+	// ExtraProgramPages specifies the additional app program len requested in pages.
+	// A page is MaxAppProgramLen bytes. This field enables execution of app programs
+	// larger than the default config, MaxAppProgramLen.
+	ExtraProgramPages int `codec:"apepp"`
```
Can we make this "apep"?  All the other msgpack fields for app data are apXX.

jannotti reviewed (commented):
Look for: 
func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusParams) error {

in data/transactions/transaction.go.  We ought to complain if extra pages is set for anything besides an app create txn.

algorandskiy reviewed (commented):
still need more unit tests:
1. pre and post protocol update
2. non-zero `ExtraProgramPages` for a regular (non-create) app call
3. total pages and app deletion
4. min balance test for old and new calculations?

shiqizng commented:
> still need more unit tests:
> 
> 1. pre and post protocol update
> 2. non-zero `ExtraProgramPages` for a regular (non-create) app call
> 3. total pages and app deletion
> 4. min balance test for old and new calculations?

@algorandskiy  I added unit tests for 1 and 2 as you suggested. I'm not sure it's a good idea to test these as unit tests though. It seems that what I added in transactions_test.go wouldn't make much sense anymore once future consensus is promoted to current.

algorandskiy commented:
> @algorandskiy I added unit tests for 1 and 2 as you suggested. I'm not sure it's a good idea to test these as unit tests though. It seems that what I added in transactions_test.go wouldn't make much sense anymore once future consensus is promoted to current.

This is easy to fix. Instead of
```
curProto := config.Consensus[protocol.ConsensusCurrentVersion]
```
use
```
curProto := config.Consensus[protocol.ConsensusV27]
```

algorandskiy commented on data/transactions/transaction_test.go:
```diff
@@ -159,6 +259,69 @@ func TestWellFormedErrors(t *testing.T) {
+				},
+			},
+			spec:          specialAddr,
+			proto:         curProto,
```
this test will break after protocol upgrade

shiqizng replied:
should I remove this?

algorandskiy replied:
no, just use v27 proto, not current.

algorandskiy commented on test/e2e-go/features/transactions/accountv2_test.go:
```diff
@@ -154,7 +154,7 @@ int 1
 	tx, err := client.MakeUnsignedAppCreateTx(
 		transactions.OptInOC, approvalOps.Program, clearstateOps.Program, schema, schema, nil, nil, nil, nil,
-	)
+		0)
```
weird formatting here and below - move 0 to the end of args list

algorandskiy commented on data/transactions/logic/eval_test.go:
```diff
@@ -1458,6 +1458,17 @@ int 2
+
+const testTxnProgramTextV4 = testTxnProgramTextV3 + `
+assert
+txn AppProgramExtraPages
```
please add `gtxn` test as well

algorandskiy reviewed (commented):
Overall looks good. Please move `MaxAvailableAppProgramLen` out of consensus params as JJ suggested. See additional comments below.

jannotti commented on data/transactions/transaction.go:
```diff
@@ -382,12 +385,16 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 
-		if len(tx.ClearStateProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("clear state program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if uint32(len(tx.ClearStateProgram)) > (1+tx.ExtraProgramPages)*uint32(proto.MaxAppProgramLen) {
+			return fmt.Errorf("clear state program too long. max len %d bytes", (1+tx.ExtraProgramPages)*uint32(proto.MaxAppProgramLen))
```
I'd like to see the formatting of this be identical to the formatting of the ApprovalProgram check.

jannotti commented on data/transactions/transaction_test.go:
```diff
@@ -121,10 +121,111 @@ func TestGoNonparticipatingWellFormed(t *testing.T) {
+					FirstValid: 100,
+				},
+				ApplicationCallTxnFields: ApplicationCallTxnFields{
+					ApplicationID: 0,
+					ApplicationArgs: [][]byte{
+						[]byte("write"),
+					},
+					ExtraProgramPages: 3,
```
To test extra pages, don't you need to try inserting a big program?  I'm somewhat surprised this is considered well formed at all, since it's trying to create an app without any program bytes. But maybe the empty program is allowed?

shiqizng replied:
I believe this is done in userBalance_test.go

jannotti replied:
That only confirms that we can _encode_ accounts with large apps in them.  So we know the max account size is still bounded by MaxEncodedAccountDataSize.  But this doesn't test that we actually _accept_ the txn. So the code where you compare the incoming program size to the value determined by extra pages is never exercised.

I'm not even sure if we've tested your allocbounds work, to ensure we can encode txns with very large programs in them.

jannotti commented on test/e2e-go/cli/goal/expect/statefulTealCreateAppTest.exp:
```diff
@@ -114,6 +114,17 @@ proc statefulTealTest { TEST_ALGO_DIR TEST_DATA_DIR TEAL_PROGRAM} {
     ::AlgorandGoal::AppDelete $APP_ID $WALLET_1_NAME $WALLET_1_PASSWORD $ACCOUNT_1_ADDRESS "str:hello" $TEST_PRIMARY_NODE_DIR
 
+    # App create with extra pages option should succeed
+    ::AlgorandGoal::AppCreateExPages $PRIMARY_WALLET_NAME "" $PRIMARY_ACCOUNT_ADDRESS ${TEAL_PROGS_DIR}/${TEAL_PROGRAM} "str:hello" $GLOBAL_BYTE_SLICES $LOCAL_BYTE_SLICES ${TEAL_PROGS_DIR}/clear_program_state.teal $TEST_PRIMARY_NODE_DIR 2
```
Again, we don't want to just test that extra pages value is set, but that because it's set specifying a large program is allowed. So you need to feed it a big approval program.  (Which may require changing goal to allow such a long program when the proto allows extra-pages)

shiqizng replied:
what's the simplest way to generate a large teal file? or is there one I could use?

jannotti replied:
Since this command is expecting teal source code, you can create a file that starts "#pragma version 3" and then has as many lines of "int 1" as you'd like. Each "int 1" will contribute one more byte to the program length.

This app won't return success is executed (because the stack may overflow, and if it doesn't it still won't end properly) but that won't matter for your tests.

jannotti replied:
I have some code in tealConsensusTest.exp that creates long teal programs.

jannotti reviewed (changes requested):
I think we need to have the tests test whether big programs are actually allowed, unless I misunderstood the tests.  They appear to only test whether the extra-pages option is allowed to be set.
